plugins/inputs/firehose/firehose_request.go

```diff
@@ -0,0 +1,156 @@
+package firehose
+
+import (
+	"compress/gzip"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/influxdata/telegraf/config"
+)
+
+type firehoseRecord struct {
+	EncodedData string `json:"data"`
+}
+
+type firehoseRequestBody struct {
+	RequestID string           `json:"requestId"`
+	Timestamp int64            `json:"timestamp"`
+	Records   []firehoseRecord `json:"records"`
+}
+
+// https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html#requestformat
+type firehoseRequest struct {
+	req                *http.Request
+	body               firehoseRequestBody
+	responseStatusCode int
+}
+
+func (r *firehoseRequest) authenticate(expectedAccessKey config.Secret) error {
+	if expectedAccessKey.Empty() {
+		return nil
+	}
+	accessKey, err := expectedAccessKey.Get()
+	if err != nil {
+		return fmt.Errorf("getting accesskey failed: %w", err)
+	}
+	reqAccessKey := r.req.Header.Get("x-amz-firehose-access-key")
+	if reqAccessKey != accessKey.String() {
+		r.responseStatusCode = http.StatusUnauthorized
+		return fmt.Errorf("unauthorized request %s", r.req.Header.Get("x-amz-firehose-request-id"))
+	}
+	return nil
+}
+
+func (r *firehoseRequest) validate() error {
+	requestID := r.req.Header.Get("x-amz-firehose-request-id")
+	if requestID == "" {
+		r.responseStatusCode = http.StatusBadRequest
+		return errors.New("x-amz-firehose-request-id header is not set")
+	}
+
+	// Check if content length is not over 64 MB.
+	if r.req.ContentLength > int64(64*1024*1024) {
+		r.responseStatusCode = http.StatusRequestEntityTooLarge
+		return fmt.Errorf("content length too large in request %s", requestID)
+	}
+
+	// Check if the requested HTTP method is allowed.
+	isAcceptedMethod := false
+	for _, method := range allowedMethods {
+		if r.req.Method == method {
+			isAcceptedMethod = true
+			break
+		}
+	}
+	if !isAcceptedMethod {
+		r.responseStatusCode = http.StatusMethodNotAllowed
+		return fmt.Errorf("forbidden method, %s, in request %s", r.req.Method, requestID)
+	}
+
+	contentType := r.req.Header.Get("content-type")
+	if contentType != "application/json" {
+		r.responseStatusCode = http.StatusBadRequest
+		return fmt.Errorf("unaccepted content type, %s, in request %s", contentType, requestID)
+	}
+
+	contentEncoding := r.req.Header.Get("content-encoding")
+	if contentEncoding != "" && contentEncoding != "gzip" {
+		r.responseStatusCode = http.StatusBadRequest
+		return fmt.Errorf("unaccepted content encoding, %s, in request %s", contentEncoding, requestID)
+	}
+
+	err := r.extractBody()
+	if err != nil {
+		return err
+	}
```

Comment:
I would combine those and not move it to an own function as it hampers the reading flow
```suggestion
	encoding := r.req.Header.Get("content-encoding")
	body, err := internal.NewStreamContentDecoder(encoding, r.req.Body)
	if err != nil {
		return fmt.Errorf("creating decoder for %q failed: %w", encoding, err)
	}
	defer r.req.Body.Close()
	
	if err := json.NewDecoder(body).Decode(&r.body); err != nil {
		r.responseStatusCode = http.StatusBadRequest
		return fmt.Errorf("decode gzip body failed: %w", err)
	}
```

Reply:
Noted. Atm, firehose can only be configured to use gzip as per the [documentation](https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html). Do you think it'd be good to keep the gzip check which I already added?
```
if contentEncoding != "" && contentEncoding != "gzip" {
		r.responseStatusCode = http.StatusBadRequest
		return fmt.Errorf("unaccepted content encoding, %s, in request %s", contentEncoding, requestID)
	}
```

Reply:
Nvm. Figured we only use the gzip decoder in the `NewStreamContentDecoder` method

Reply:
done